servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ConcurrentStack.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright © 2020 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Consumer;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.internal.ThrowableUtils.catchUnexpected;
+import static io.servicetalk.utils.internal.PlatformDependent.throwException;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicReferenceFieldUpdater.newUpdater;
+
+final class ConcurrentStack<T> {
+    private static final Node<?> CLOSED = new Node<>();
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<ConcurrentStack, Node> topUpdater =
+            newUpdater(ConcurrentStack.class, Node.class, "top");
+    @Nullable
+    private volatile Node<T> top;
+
+    boolean push(T item) {
+        final Node<T> newTop = new Node<>(item);
+        Node<T> oldTop;
+        do {
+            oldTop = top;
+            if (oldTop == CLOSED) {
+                return false;
+            }
+            newTop.next = oldTop;
```

Comment:
I don't think it is worth the complexity as this will only help in specific scenarios and we are assuming small sets. I think if we run into issues with the "relaxed" reclamation policy we can provide an alternative implementation which is more strict.

Reply:
sgtm!